### PR TITLE
CDAP-6378 Sort jars to be added to the MapReduce classpath so that CDAP classes get picked up before Twill.

### PR DESCRIPTION
**Pull request by sagarkapare**
JIRA: https://issues.cask.co/browse/CDAP-6378

### CODE REVIEW

**Comment by anwar6953** [comment]
We didn't already have this?

**Comment by sagarkapare** [comment]
nope.

**Comment by anwar6953** [comment]
Oh, we had this for things other than MapReduce (i.e. master service and containers other than MR)?

**Comment by sagarkapare** [comment]
It was done for just master:
https://github.com/caskdata/cdap/pull/5040

**Comment by sagarkapare** [comment]
Build is running here - http://builds.cask.co/browse/CDAP-DUT4328-1

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java [hunk 1]
```diff
@@ -257,11 +260,25 @@ protected void startUp() throws Exception {
           classpath.add(logbackLocation.getName());
         }
 
-        // Generate and localize the launcher jar to control the classloader of MapReduce containers processes
-        classpath.add("job.jar/lib/*");
+        // Get all the jars in jobJar and sort them lexically before adding to the classpath
+        // This allows CDAP classes to be picked up first before the Twill classes
+        JarFile jobJarFile = new JarFile(jobJar);
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java]: Use try resource block to make sure the jar file is closed

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java [hunk 2]
```diff
@@ -257,11 +260,25 @@ protected void startUp() throws Exception {
...
+        Enumeration<JarEntry> entries = jobJarFile.entries();
+        List<String> jarFiles = new ArrayList<>();
+        while (entries.hasMoreElements()) {
+          JarEntry entry = entries.nextElement();
+          if (entry.getName().endsWith(".jar")) {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java]: This is not the same as the existing logic. We should only include jars under the lib dir

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java (hunk 1 above)
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java]: fixed.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java (hunk 2 above)
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java]: yes. right fixed the code.

**Comment by sagarkapare** [comment]
@chtyim Addressed comments. Please take a look again. Thanks!

**Comment by chtyim** [comment]
LGTM

**Comment by sagarkapare** [comment]
Checkstyle passes. Merging PR before next ITN.

**Pull request merged**